Comments Redesign: Add the actions

Copons opened the pull request:
Follow up to: #19280

Add the Approve/Unapprove, Spam, Trash, and Like actions to the new `Comment` component.
Moved the whole logic away from `CommentList` and inside `CommentActions`, as close as possible to the various action buttons.

This implementation is barebone and lacks some features:

- No Edit and Reply actions.
- No `CommentList` cleanup.
- No success notices and therefore **no undos**.

### Notices and Undos

Internal discussion: p7jreA-1ca-p2

We've been talking of dropping success notices altogether. The only thing blocking us is that the notices are currently necessary for the undos.

@vindl suggested an approach similar to the one already featured in the Posts List.
On spam/trash (which are the "semi-destructive" actions), the comment persist and enter a "disabled" state. It gets covered by an overlay with a message such as "Comment Deleted - Undo?".

By moving the Undo into the actual comment, we'd drastically reduce all the required back and forth between `Comment` and `CommentList`.
Basically we'd end up with just the persistence functions, which can't go anywhere else but in the `List`.
There will be a little bit of duplication between the single and the bulk actions, but it's a fair trade-off imho.

What about approve/unapprove? I'd actually suggest that we don't persist them anymore.
We have the All list where they're shown together, and I can't see any reason why the Approved and Pending lists should behave _almost but not completely_ the same.

### Screenshots (WIP)

Large screens collapsed
<img width="737" alt="screen shot 2017-11-01 at 10 30 51" src="https://user-images.githubusercontent.com/2070010/32270985-e6a06dc8-beef-11e7-8ed7-002971291d22.png">

Large screens expanded
<img width="738" alt="screen shot 2017-11-01 at 10 31 14" src="https://user-images.githubusercontent.com/2070010/32270986-e6bba0ac-beef-11e7-940d-745d3a9f11c4.png">

Small screens collapsed
<img width="447" alt="screen shot 2017-11-01 at 10 33 03" src="https://user-images.githubusercontent.com/2070010/32271055-2f4cc620-bef0-11e7-9027-fa09b53b3f94.png">

Small screens expanded
<img width="448" alt="screen shot 2017-11-01 at 10 33 11" src="https://user-images.githubusercontent.com/2070010/32271066-34d6f066-bef0-11e7-834e-9bbfc1e0c267.png">

### Testing instructions

Checkout this branch and run it with
`env ENABLE_FEATURES=comments/management/m3-design npm start`

gwwar commented:
For folks testing you'll want to start with `ENABLE_FEATURES=comments/management/m3-design npm start`

gwwar commented:
Found a corner case for persistence:

- Visit the All Tab
- unapprove a comment
- then click on spam
- The comment with stay instead of moving to spam

This might be a made moot by your other PR, but wanted to note it.

gwwar commented:
Actions all work for me, though some other things to note that we can follow up on:

- There's a dividing line missing from the actions in the expanded mode
- We should hide individual actions in bulk mode
- The actions without the bottom labels in mobile (unexpanded) feel kind of off since we're not saving that much space, and folks end up guessing what the actions do.
- I'm just throwing this out there, but as our cards get larger is there still much value for our unexpanded Comment view, besides truncating very large comments and for bulk mode?

Copons commented:
> Found a corner case for persistence

You're right, I totally forgot to remove comments from the persisted list (I bring the function from `List` to `Actions` but never use it).

> There's a dividing line missing from the actions in the expanded mode

Good catch, completely missed it.

> we're not saving that much space

> is there still much value for our unexpanded Comment view?

This is another thing I missed: only the pending comments have the action visible in the collapsed view. Actions for approved, spam, and trash comments only show up when expanded.
This way the difference is more evident and imho there is a value.

kwight commented:
I noticed that in bulk edit, the Pending comments don't keep their yellow background.

On smaller screens, the action buttons feel really... heavy? Dominant? I find I get the comment text lost in the big icons (anyone else feel that way?). I admit I prefer the weight given to actions on the post list; things are smaller, and seem more clearly separated or defined because of their borders. Of course, posts also take up a giant amount of screen space. Comments are hard :/

<img width="907" alt="screen shot 2017-10-31 at 4 17 15 pm" src="https://user-images.githubusercontent.com/349751/32253212-04aab4de-be57-11e7-8694-357849e7c3e7.png">

Copons commented:
Thanks for the review!
I've addressed your observations:

- Fix spam and trash actions not removing the comment from the persistent list (although, it will become obsolete if we adopt the new notices approach).

- Fix actions render conditions. Now it only loads if:
  - The comment is expanded.
  - The comment is collapsed, but only if with status `unapproved` and not in bulk mode.

- Polished the style, by adding more padding on the buttons, and adding the missing top border on small screens' expanded view.

Also updated the screenshots in the PR description.

kwight commented:
These changes feel great, @Copons ! Two small notes:

* Is the border above the actions supposed to extend the entire width of the comment card? It looks a little odd compared to the border above it that does.
* Also noticed that we can no longer click anywhere on the comment when in bulk mode to select it; we need to click the checkbox now. I think it's better the other way?

Copons commented:
> Is the border above the actions supposed to extend the entire width of the comment card? It looks a little odd compared to the border above it that does.

I've followed @megsfulton's proposal, but you make a good point comparing it to the post actions.
Though, I think we could ship this as is and discuss+iterate later while it's all still hidden.

> Also noticed that we can no longer click anywhere on the comment when in bulk mode to select it; we need to click the checkbox now. I think it's better the other way?

Yup, it's something I noticed already and I'll push a fix asap. 🙂

Copons commented:
Entire card clickable on bulk mode taken care in https://github.com/Automattic/wp-calypso/pull/19404

Copons commented:
PR rebased, now it shows a complete `Comment` card, with both actions and reply textarea.

If @vindl and @drw158 want to check out how it all looks like with all comments expanded, https://github.com/Automattic/wp-calypso/pull/19332/commits/5c4d45aa8a7da7c52c0029fe4e26c4dfbf3e8026 is a good commit to branch off.
